Add SingleStore vectorstore integration

volodymyr-memsql opened the pull request:
# SingleStore vector store

* Added SingleStoreVectorStore class that allows using SingleStoreDB as a vector store. 
* Added integration test
* Added documentation

## Future work 
* Enable filtering by metadata in `similaritySearch`

volodymyr-memsql commented:
Hello, @jacoblee93! I cannot assign a reviewer to the PR. What is the correct procedure to get this reviewed and merged?  Thanks!

jacoblee93 commented:
Hey @volodymyr-memsql, looking now!

jacoblee93 commented on package.json:
```diff
@@ -43,13 +43,15 @@
     "typescript": "^5.0.0"
   },
   "dependencies": {
+    "mysql": "^2.18.1",
```
Shouldn't be a hard dependency here. Please see this guide for the right way to add third-party deps:
https://github.com/hwchase17/langchainjs/blob/main/.github/contributing/INTEGRATIONS.md

jacoblee93 commented on langchain/src/vectorstores/singlestore.ts:
```diff
@@ -0,0 +1,118 @@
+    return new Promise((resolve, reject) => {
+      this.connectionPool.getConnection((err: MysqlError, connection: PoolConnection) => {
+        if (err) reject(err);
+        connection.query(searchQuery, (err: MysqlError, rows: any[]) => {
```
Try to avoid `any`

jacoblee93 commented on langchain/src/vectorstores/singlestore.ts:
```diff
@@ -0,0 +1,118 @@
+    // use vector DOT_PRODUCT as a distance function 
+    const searchQuery = `SELECT ${this.contentColumnName}, ${this.metadataColumnName}, DOT_PRODUCT(${this.vectorColumnName}, JSON_ARRAY_PACK('[${query}]')) as __score  FROM ${this.tableName} ORDER BY __score DESC LIMIT ${k}`;
+    return new Promise((resolve, reject) => {
+      this.connectionPool.getConnection((err: MysqlError, connection: PoolConnection) => {
```
Is there no `async` method on `connectionPool`? Prefer not to mix callbacks and promises.

If there isn't, is this the right package to use? Last publish was 3 years ago, seems abandoned.

jacoblee93 reviewed (changes requested):
Thank you, and apologies for the delay. Left a few comments.

jacoblee93 commented:
Also, please remember to `yarn lint` and `yarn format`!

volodymyr-memsql commented:
Thanks for the feedback, @jacoblee93! I switched to the connection pool from `mysql2/promise` and got rid of callbacks. Also, addressed other comments. Could you please check it again?

jacoblee93 commented on langchain/src/vectorstores/singlestore.ts:
```diff
@@ -0,0 +1,135 @@
+        try {
+          await this.connectionPool.execute(
+            `INSERT INTO ${tableName} VALUES ('${
+              documents[idx].pageContent
```
I don't think we can assume documents are friendly/sanitized - can you use a prepared statement here?

jacoblee93 replied:
It's just good practice regardless

jacoblee93 commented on langchain/src/vectorstores/index.ts:
```diff
@@ -8,3 +8,4 @@ export { PineconeStore } from "./pinecone.js";
 export { VectorStore, SaveableVectorStore } from "./base.js";
 export { SupabaseVectorStore } from "./supabase.js";
 export { PrismaVectorStore } from "./prisma.js";
+export { SingleStoreVectorStore } from "./singlestore.js";
```
Exporting from here is deprecated, please remove.

jacoblee93 commented on langchain/src/vectorstores/singlestore.ts:
```diff
@@ -0,0 +1,135 @@
+    ] = await this.connectionPool.query(
+      `SELECT ${this.contentColumnName},
+    ${this.metadataColumnName},
+    DOT_PRODUCT(${this.vectorColumnName}, JSON_ARRAY_PACK('[${query}]')) as __score
```
I feel like this should be ok but best practices is still to parameterize

jacoblee93 reviewed (changes requested):
A few more changes/questions - potential SQL injection

jacoblee93 commented on examples/src/indexes/vector_stores/singlestore.ts:
```diff
@@ -0,0 +1,26 @@
+    user: process.env.SINGLESTORE_USERNAME,
+    password: process.env.SINGLESTORE_PASSWORD,
+    database: process.env.SINGLESTORE_DATABASE,
+  });
```
I think generally I'd have a preference for making `mysql2` an implementation detail within your store - basically passing in `connectionParams` rather than passing in the entire `connectionPool`. Fewer imports for the developer, and allows you to swap packages more seamlessly if something should happen that dependency. 

But up to you! Looks good otherwise and will merge by tomorrow with or without that change.

volodymyr-memsql replied:
The `createPool` allows a developer to pass many different parameters (not only those four that I used in the example) depending on the type of connection. I'd prefer to keep this flexibility and not copy-paste all parameters as parameters of the vector store.

jacoblee93 replied:
They may export a type like `ConnectionPoolParams` with all options that you can then initialize a pool with. No worries though, will merge shortly.

jacoblee93 reviewed (approved):
Looks good, one small suggestion Will merge by tomorrow with or without that change.
